### DISCUSSION TITLE
Layout problem

### DISCUSSION THREAD

**Discussion by mharig** [Textualize/textual#1844] [Q&A] [answered]
I want a 3x3 grid with a center widget that is wider than the neighbour widgets (no row-span!). All widgets are TextLog, except for the center widget, that consists of a top TextLog, and bottom Input, the latter with height 1.
I achieved all of the above, but to get the two widgets of the center 'cell' displayed, I have to set the heights of the parent, the TextLog and the Input explicitly to the appropriate heights. This is not portable to other screen sizes, so I definitely want an automatic layout.
I didn't provide code, because I do not want to push someone in the obvious not-quiet-right direction I took.

**Comment by davep** [comment, marked as answer]
Something like this?

![Screenshot 2023-02-20 at 14 39 22](https://user-images.githubusercontent.com/28237/220136073-88732a4a-e87c-49e2-9254-f3526d0924dc.png)

```python
from textual.app        import App, ComposeResult
from textual.containers import Horizontal, Vertical
from textual.widgets    import Header, Footer, TextLog, Input

class NotAGridLayout( App[ None ] ):

    CSS = """
    .left, .middle, .right {
        border: round #777;
    }

    .left, .right {
        width: 1fr;
    }

    .middle {
        width: 2fr;
    }

    TextLog {
        height: 1fr;
    }
    """

    def compose( self ) -> ComposeResult:
        yield Header()
        yield Vertical(
            Horizontal( TextLog( classes="left" ), TextLog( classes="middle" ), TextLog( classes="right" ) ),
            Horizontal(
                TextLog( classes="left" ),
                Vertical( TextLog(), Input( placeholder="Here's the Input" ), classes="middle" ),
                TextLog( classes="right" )
            ),
            Horizontal( TextLog( classes="left" ), TextLog( classes="middle" ), TextLog( classes="right" ) ),
        )
        yield Footer()

if __name__ == "__main__":
    NotAGridLayout().run()
```

**Reply by mharig** [reply]
Not exactly, the center 'cell' should be wider, not the center column, so I used a grid layout for the middle row with grid-size 1fr 2fr 1fr.
However, that is near identical to my approach, but it does not work. I will use your approach and change it in the direction I want and see what will happen. Thanks so far!

**Reply by davep** [reply]
So more like this?

![Screenshot 2023-02-20 at 14 53 56](https://user-images.githubusercontent.com/28237/220139268-9db4b56d-bb94-439c-b1b7-875a9ed40a4c.png)

```python
from textual.app        import App, ComposeResult
from textual.containers import Horizontal, Vertical
from textual.widgets    import Header, Footer, TextLog, Input

class NotAGridLayout( App[ None ] ):

    CSS = """
    Horizontal > * {
        border: round #777;
    }

    Horizontal > Vertical {
        width: 2fr;
    }

    TextLog {
        height: 1fr;
        width: 1fr;
    }
    """

    def compose( self ) -> ComposeResult:
        yield Header()
        yield Vertical(
            Horizontal( TextLog(), TextLog(), TextLog() ),
            Horizontal(
                TextLog(),
                Vertical( TextLog(), Input( placeholder="Here's the Input" ) ),
                TextLog()
            ),
            Horizontal( TextLog(), TextLog(), TextLog() ),
        )
        yield Footer()

if __name__ == "__main__":
    NotAGridLayout().run()
```

**Reply by mharig** [reply]
Yes. I now got it how I want by just changing the class of the Vertical with the Input to a class with width: 2fr.
What I do not know is: why does my approach not work. I use for the 3 Horizontal rows a grid layout with grid-size: 3 1  and for the middle row grid-columns: 1fr 2fr 1fr. The rest is near exactly like your approach. But anyhow, my problem is solved. Thanks a lot!

**Reply by davep** [reply]
Without actually seeing the code it's hard to say what the issue might be.

**Reply by mharig** [reply]
Of course. The following is a mix of your approach (mostly the css) and my approach, and it does not work either.
```python
from textual.app import App
from textual.widgets import TextLog, Header, Footer, Input
from textual.containers import Vertical, Horizontal, Container


class BTRui(App):
    CSS_PATH = 'BTRui.css'
    
    def compose(self):
        yield Header(show_clock=True)
        self.tllog = TextLog(highlight=True, markup=True, wrap=True, id='tllog', classes='left')
        self.tclog = TextLog(highlight=True, markup=True, wrap=True, id='tclog', classes='middle')
        self.trlog = TextLog(highlight=True, markup=True, wrap=True, id='trlog', classes='right')
        self.cllog = TextLog(highlight=True, markup=True, wrap=True, id='cllog', classes='left')
        self.shellog = TextLog(highlight=True, markup=True, wrap=True, id='shellog')
        self.cmdline = Input(placeholder='Type command', id='cmdline')
        self.cclog = Vertical(self.shellog, self.cmdline, classes='shellog')
        self.crlog = TextLog(highlight=True, markup=True, wrap=True, id='crlog', classes='right')
        self.bllog = TextLog(highlight=True, markup=True, wrap=True, id='bllog', classes='left')
        self.bclog = TextLog(highlight=True, markup=True, wrap=True, id='bclog', classes='middle')
        self.brlog = TextLog(highlight=True, markup=True, wrap=True, id='brlog', classes='right')
        yield Vertical(Horizontal(self.tllog,self.tclog,self.trlog),
                       Horizontal(self.cllog,self.cclog,self.crlog),
                       Horizontal(self.bllog,self.bclog,self.brlog))


    def on_ready(self):
        self.tllog.write('topleft')
        self.tclog.write('topcenter')
        self.trlog.write('topright')
        self.cllog.write('centerleft')
        self.shellog.write('shell output')
        self.shellog.write('01234567890123456789012345678901234567890123456789012345678901234567890123456789')
        self.cmdline.focus()
        self.crlog.write('centerright')
        self.crlog.write('01234567890123456789012345678901234567890123456789012345678901234567890123456789')
        self.bllog.write('bottomleft')
        self.bclog.write('bottomcenter')
        self.brlog.write('bottomright')


if __name__ == '__main__':
    app = BTRui()
    app.run()

```

BTRui.css:

```css
.left, .middle, .right {
    border: round #777;
}

.left, .right {
    width: 1fr;
}

.shellog {
    width: 2fr;
    overflow-y: hidden;
}

#shellog {
    background: blue;
    border: none;
}

#cmdline {
    background: red;
    border: none;
}
```

**Reply by davep** [reply]
The problem seems to be that you've sized `.left` and `.right` to be `1fr` but left `.middle` as its default size. I suspect you want to change this:

```scss
.left, .right {
    width: 1fr;
}
```

to this:

```scss
.left, .middle, .right {
    width: 1fr;
}
```

**Reply by mharig** [reply]
The thing that does not work is: the Input field is not displayed.

**Reply by davep** [reply]
In that case you've got a couple of things going on in your code above. So first, the problem I highlighted:

![Screenshot 2023-02-21 at 21 26 33](https://user-images.githubusercontent.com/28237/220462249-2c262b78-4a05-42d7-9042-0d5cf3827993.png)

Notice how the left and right panes on the top and bottom aren't really showing? The fix I mention above fixes that:

![Screenshot 2023-02-21 at 21 28 08](https://user-images.githubusercontent.com/28237/220462451-37526cd2-f204-4620-8815-636afc8c286f.png)

As for your input: you've not told the `#shellog` to share space with the input in the container, so I'd also add `height: 1fr;` to the styles for `#shelllog`. That results in this:

![Screenshot 2023-02-21 at 21 31 24](https://user-images.githubusercontent.com/28237/220463046-5faa4b54-c819-42d7-8620-5c97eefcae80.png)

**Reply by mharig** [reply]
Fits perfectly!
I still do not understand why it is necessary to tell shellog explixitly to share space. Shouldn't that be an issue of the parent container, i.e. Vertical?

**Reply by davep** [reply]
Not really, no. I guess the way to think of it is like this (it's not a perfect analogy, but it roughly works). If you were packing physical things in a box, the box you're putting stuff in doesn't control the size, you pick the right things to fit inside it. So... the container holds your widgets and controls where they end up, but to do that they have to say "I'm this big!".
